File: 9-error_handling.py
# Python 用 try/except/else/finally 分清成功、失敗同收尾動作。
def parse_age(value):
    try:
        age = int(value)
    except ValueError:
        return "Invalid age"
    else:
        return f"Age: {age}"
    finally:
        print(f"Finished parsing {value}")


print(parse_age("42"))
print("---")
print(parse_age("abc"))

print("---")


# 自訂 Exception class 可以令 except 捕捉範圍清楚啲。
class PaymentError(Exception):
    pass


def charge(amount):
    if amount <= 0:
        raise PaymentError("Amount must be positive")

    return f"Charged ${amount}"


try:
    print(charge(0))
except PaymentError as error:
    print(f"Payment failed: {error}")

File: 9-error_handling.rb
# Ruby 用 rescue/ensure 處理例外；ensure 無論成功失敗都會執行。
def parse_age(value)
  age = Integer(value)
  "Age: #{age}"
rescue ArgumentError
  'Invalid age'
ensure
  puts "Finished parsing #{value}"
end

puts parse_age('42')
puts '---'
puts parse_age('abc')

puts '---'

# 自訂 error class 可以令呼叫端 rescue 得更精準。
class PaymentError < StandardError
end

def charge(amount)
  raise PaymentError, 'Amount must be positive' if amount <= 0

  "Charged $#{amount}"
end

begin
  puts charge(0)
rescue PaymentError => e
  puts "Payment failed: #{e.message}"
end
